File: 3819-CountCoveredBuildings/3819-CountCoveredBuildings.cpp
// Last updated: 08/03/2026, 13:22:58
#define vi vector<int>

class Solution {
public:

    int countCoveredBuildings(int n, vector<vector<int>>& buildings) {
        n ++;
        vi leftMostBuildingOnY(n, n+1);
        vi rightMostBuildingOnY(n, -1);
        vi lowestBuildingOnX(n, n+1);
        vi highestBuildingOnX(n, 0);

        int res = 0;

        for(auto building: buildings) {
            int x = building[0];
            int y = building[1];

            leftMostBuildingOnY[y] = min(x, leftMostBuildingOnY[y]);
            rightMostBuildingOnY[y] = max(x, rightMostBuildingOnY[y]);

            lowestBuildingOnX[x] = min(y, lowestBuildingOnX[x]);
            highestBuildingOnX[x] = max(y, highestBuildingOnX[x]);
        }

        for(auto building: buildings) {
            int x = building[0];
            int y = building[1];

            if (y > lowestBuildingOnX[x] && y < highestBuildingOnX[x] 
            && x > leftMostBuildingOnY[y] && x < rightMostBuildingOnY[y]) {
                res ++;
            }


        }

        return res;
    }
};

File: 3819-CountCoveredBuildings/3819-CountCoveredBuildings.java
// Last updated: 08/03/2026, 13:23:02
class Solution {
    public int countCoveredBuildings(int n, int[][] buildings) {
        Map<Integer,TreeSet<Integer>> colToRows=new HashMap<>();
        Map<Integer,TreeSet<Integer>> rowToCols=new HashMap<>();
        
        for(int[] b:buildings){
            colToRows.computeIfAbsent(b[1], k -> new TreeSet<>()).add(b[0]);
            rowToCols.computeIfAbsent(b[0], k -> new TreeSet<>()).add(b[1]);
        }
        
        int cnt=0;
        
        for(int[] b:buildings){
            
            TreeSet<Integer> xCoords=colToRows.get(b[1]), yCoords = rowToCols.get(b[0]);
            if(xCoords.lower(b[0])!=null && xCoords.higher(b[0]) != null && 
                yCoords.lower(b[1])!=null && yCoords.higher(b[1]) != null) {
                cnt++;
            }
        }
        
        return cnt;
    }
}